### PR TITLE
Let `LowerLinalgMicrokernels` assume that dynamic dims are contiguous.

### PR DESCRIPTION
**Pull request by bjacob**
This is part 2/9 of #11755: Finish the transition to `--iree-flow-enable-data-tiling`.

In the VMVX backend, when deciding whether an op can be rewritten as a microkernel call, we need to check if memref layouts inner dimeensions are contiguous row-major as expected by the microkernel.

When a dimension has dynamic size, that is hard to do as a static analysis. This PR makes us just assume that dynamic dims are contiguous, which is safe-ish for e2e workloads because bufferization only creates contiguous buffers, but is not 100% satisfactory, technically this pass in isolation is not correct on all possible inputs.

Issue #11633 tracks finding a better solution.

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp [hunk 1]
```diff
@@ -28,6 +29,22 @@ namespace iree_compiler {
 
 namespace {
 
+static llvm::cl::opt<bool> clAssumeDynamicDimsContiguous(
+    "iree-vmvx-microkernels-assume-dynamic-dims-contiguous",
+    llvm::cl::desc(
+        "In the VMVX backend, when deciding whether an op can be rewritten as "
+        "a microkernel call, we need to check if memref layouts inner "
+        "dimeensions are contiguous row-major as expected by the microkernel. "
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp:37]: dimensions*

**Review by benvanik — CHANGES_REQUESTED** [review]
I'm really skeptical of the value of this - can you dumb down the description a bit and explain why you think this is needed? AFAIU it's never possible under any circumstances for there to be non-contiguous inner dimensions and I really don't like the idea of this flag implying otherwise and defaulting to false.

File: compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp:37]: What do you mean -- it's easy enough to construct a memref with non-contiguous inner dims, just take a memref.subview out of a larger memref. Here the memrefs we are interested in are subviews generated by the tile-and-distribute-to-workgroups pass. That they happen to be contiguous depends on the actual runtime values of SSA values controlling their layout. Issue #11633 referenced in the PR description has details, including the typical post-tiling IR in question in https://gist.github.com/bjacob/59f3ea6bf40aef7915fff9f6a287a599

Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/VMVX/LowerLinalgMicrokernels.cpp:37]: Talked with @MaheshRavishankar . What I was not getting, was that bufferization always creates contiguous buffers.  Therefore the existing contiguousness test was always futile as far as e2e workloads were concerned. We agreed that I would remove the commandline flag, and instead make it a pass option (not exposed on the compiler command line, for debugging only) and default to true.

**Review by benvanik — APPROVED** [review]

**Pull request merged**